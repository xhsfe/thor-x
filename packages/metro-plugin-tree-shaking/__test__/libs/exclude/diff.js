export const diff = (a, b) => a - b
