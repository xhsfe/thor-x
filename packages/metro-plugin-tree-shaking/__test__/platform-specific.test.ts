import { execSync } from 'child_process'
import 'colors'
import fs from 'fs'
import path from 'path'
import { diffLines } from 'diff'

test('tree shaking stats is pure, final live state of parts is decided by metroGraph in serializer', () => {
  const METRO_TREE_SHAKING_CACHE_ROOT = path.join(
    __dirname,
    '../.optimize/platform-specific',
  )
  const OUTPUT_ROOT = path.join(
    __dirname,
    '../dist/platform-specific',
  )
  const ENTRY_FILE = './__test__/platform-specific.ts'
  const BUNDLE_OUTPUT = './dist/platform-specific'

  if (fs.existsSync(OUTPUT_ROOT)) {
    fs.rmSync(OUTPUT_ROOT, { recursive: true })
  }

  fs.mkdirSync(OUTPUT_ROOT, { recursive: true })

  if (fs.existsSync(METRO_TREE_SHAKING_CACHE_ROOT)) {
    fs.rmSync(METRO_TREE_SHAKING_CACHE_ROOT, { recursive: true })
  }

  execSync(
    `PRE_BUILD=1 METRO_TREE_SHAKING_CACHE_ROOT=${METRO_TREE_SHAKING_CACHE_ROOT} react-native bundle --platform=android --entry-file=${ENTRY_FILE} --bundle-output=${BUNDLE_OUTPUT}/main.jsbundle --sourcemap-output=${BUNDLE_OUTPUT}/main.jsbundle.map --dev=true`,
  )

  execSync(
    `PRE_BUILD=1 METRO_TREE_SHAKING_CACHE_ROOT=${METRO_TREE_SHAKING_CACHE_ROOT} react-native bundle --platform=ios --entry-file=${ENTRY_FILE} --bundle-output=${BUNDLE_OUTPUT}/main.jsbundle --sourcemap-output=${BUNDLE_OUTPUT}/main.jsbundle.map --dev=true`,
  )

  execSync(
    `DISABLE_OPTIMIZE=1 METRO_TREE_SHAKING_CACHE_ROOT=${METRO_TREE_SHAKING_CACHE_ROOT} react-native bundle --platform=native --entry-file=${ENTRY_FILE} --bundle-output=${BUNDLE_OUTPUT}/origin.jsbundle --sourcemap-output=${BUNDLE_OUTPUT}/origin.jsbundle.map --dev=true`,
  )

  execSync(
    `METRO_TREE_SHAKING_CACHE_ROOT=${METRO_TREE_SHAKING_CACHE_ROOT} react-native bundle --platform=android --entry-file=${ENTRY_FILE} --bundle-output=${BUNDLE_OUTPUT}/optimize.android.jsbundle --sourcemap-output=${BUNDLE_OUTPUT}/optimize.android.jsbundle.map --dev=true`,
  )

  execSync(
    `METRO_TREE_SHAKING_CACHE_ROOT=${METRO_TREE_SHAKING_CACHE_ROOT} react-native bundle --platform=ios --entry-file=${ENTRY_FILE} --bundle-output=${BUNDLE_OUTPUT}/optimize.ios.jsbundle --sourcemap-output=${BUNDLE_OUTPUT}/optimize.ios.jsbundle.map --dev=true`,
  )

  const originContent = fs.readFileSync(
    path.join(OUTPUT_ROOT, 'origin.jsbundle'),
    { encoding: 'utf8' },
  )
  const optimizedAndroidContent = fs.readFileSync(
    path.join(OUTPUT_ROOT, 'optimize.android.jsbundle'),
    { encoding: 'utf8' },
  )
  const optimizedIOSContent = fs.readFileSync(
    path.join(OUTPUT_ROOT, 'optimize.ios.jsbundle'),
    { encoding: 'utf8' },
  )
  const androidDiffs = JSON.stringify(
    diffLines(originContent, optimizedAndroidContent, {
      ignoreWhitespace: true,
    }).filter(item => item.added || item.removed),
    null,
    '\t',
  )
  expect(androidDiffs).toMatchSnapshot()

  const iOSDiffs = JSON.stringify(
    diffLines(originContent, optimizedIOSContent, {
      ignoreWhitespace: true,
    }).filter(item => item.added || item.removed),
    null,
    '\t',
  )
  expect(iOSDiffs).toMatchSnapshot()
})
