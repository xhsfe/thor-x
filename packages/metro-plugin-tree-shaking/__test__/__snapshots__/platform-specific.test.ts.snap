// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tree shaking stats is pure, final live state of parts is decided by metroGraph in serializer 1`] = `
"[
	{
		"count": 4,
		"removed": true,
		"value": "exports.add = void 0;\\nexports.div = div;\\nexports.mul = _;\\nexports.sub = void 0;\\n"
	},
	{
		"count": 1,
		"added": true,
		"value": "exports.platform = exports.add = void 0;\\n"
	},
	{
		"count": 10,
		"removed": true,
		"value": "function div(a, b) {\\nreturn a / b;\\n}\\nvar sub = exports.sub = function sub(a, b) {\\nreturn a - b;\\n};\\nfunction _(a, b) {\\nreturn a * b;\\n}\\n},1,[],\\"__test__/libs/math.js\\");\\n"
	},
	{
		"count": 2,
		"added": true,
		"value": "var platform = exports.platform = 'android';\\n},1,[],\\"__test__/libs/math.android.js\\");\\n"
	},
	{
		"count": 1,
		"removed": true,
		"value": "//# sourceMappingURL=origin.jsbundle.map"
	},
	{
		"count": 1,
		"added": true,
		"value": "//# sourceMappingURL=optimize.android.jsbundle.map"
	}
]"
`;

exports[`tree shaking stats is pure, final live state of parts is decided by metroGraph in serializer 2`] = `
"[
	{
		"count": 4,
		"removed": true,
		"value": "exports.add = void 0;\\nexports.div = div;\\nexports.mul = _;\\nexports.sub = void 0;\\n"
	},
	{
		"count": 1,
		"added": true,
		"value": "exports.platform = exports.add = void 0;\\n"
	},
	{
		"count": 10,
		"removed": true,
		"value": "function div(a, b) {\\nreturn a / b;\\n}\\nvar sub = exports.sub = function sub(a, b) {\\nreturn a - b;\\n};\\nfunction _(a, b) {\\nreturn a * b;\\n}\\n},1,[],\\"__test__/libs/math.js\\");\\n"
	},
	{
		"count": 2,
		"added": true,
		"value": "var platform = exports.platform = 'ios';\\n},1,[],\\"__test__/libs/math.ios.js\\");\\n"
	},
	{
		"count": 1,
		"removed": true,
		"value": "//# sourceMappingURL=origin.jsbundle.map"
	},
	{
		"count": 1,
		"added": true,
		"value": "//# sourceMappingURL=optimize.ios.jsbundle.map"
	}
]"
`;
